src/Arcus.WebApi.Security.Authorization/Jwt/JwtTokenReader.cs

```diff
@@ -14,18 +14,50 @@ public class JwtTokenReader : IJwtTokenReader
             "https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration";
 
         private readonly string _applicationId;
+        private readonly TokenValidationParameters _tokenValidationParameters;
```

Comment:
If I'm not mistaken, the AudienceId is also present in the `TokenValidationParameters` , so in fact, you have 2 ways here to identify the audience (the property in `TokenValidationParametes` and the `_applicationId`).

Shouldn't the ctor that accepts the `applicationId` not instantiate a `TokenValidationParameters` instance, fill out the `ValidAudience` property and specify some defaults for the other properties ?
Then the `_applicationId`member can be removed, as this only adds confusion.

Reply:
I see that you do the above  in the `ValidateToken` method; I'd do that in the ctor, and then work with the `TokenValidationParameters` in the ValidateToken method.

Reply:
That's the idea but the default uses the signing keys from the remote endpoint which requires async.
That's why I decided to defer it and use this approach: https://github.com/arcus-azure/arcus.webapi/pull/129/files/035318cff93aac825afbd56f6c7eb101455994cd#diff-afcdc5e9bb40ea179ba752371ed1d6c8R71

Reply:
Ok , I see.

But then, why not use a Factory method instead of this constructor:

```csharp
public static async Task<JwtTokenReader> CreateAsync( string applicationId )
{
     OpenIdConnectConfiguration config = await _configManager.GetConfigurationAsync();

     var validationParameters = new TokenValidationParameters	 
     {	            
         ValidateAudience = true,
         ValidAudience = applicationId,	 
         ValidateIssuer = false,	                
         IssuerSigningKeys = config.SigningKeys,
         ValidateLifetime = true	                
     };

     return new JwtTokenReader(validationParameters);
}
```
something like this.

Just my 2 cents.  Then offcourse, this is not very much inline with the other options to instantiate a JwtTokenReader, since there, constructors are used.

Anyway, since `_applicationId` is a private member, and this is just internal workings, I can agree with how it is implemented.

/writing-while-thinking

Reply:
That's certainly another approach but problem is that startup is not async as well so you'll have the same issue unless there is a way which I don't know of course!

Reply:
Discussed this with Frederik and he's ok with the current approach although it's not ideal, but nothing we can do about it